// @panter/vue-i18next/lib/wait.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.bind = bind;
exports.update = update;

var _utils = require('./utils');

function assert(vnode) {
  var vm = vnode.context;

  if (!vm.$i18n) {
    (0, _utils.warn)('No VueI18Next instance found in the Vue instance');
    return false;
  }

  return true;
} /* eslint-disable no-param-reassign, no-unused-vars */

function waitForIt(el, vnode) {
  if (vnode.context.$i18n.i18next.isInitialized) {
    el.hidden = false;
  } else {
    el.hidden = true;
    var initialized = function initialized() {
      vnode.context.$forceUpdate();
      // due to emitter removing issue in i18next we need to delay remove
      setTimeout(function () {
        if (vnode.context && vnode.context.$i18n) {
          vnode.context.$i18n.i18next.off('initialized', initialized);
        }
      }, 1000);
    };
    vnode.context.$i18n.i18next.on('initialized', initialized);
  }
}

function bind(el, binding, vnode) {
  if (!assert(vnode)) {
    return;
  }

  waitForIt(el, vnode);
}

function update(el, binding, vnode, oldVNode) {
  if (vnode.context.$i18n.i18next.isInitialized) {
    el.hidden = false;
  }
}

exports.default = {
  bind: bind,
  update: update
};
